src/Microsoft.OData.Client/ALinq/ResourceBinder.cs

```diff
@@ -2910,6 +2988,46 @@ internal static void ValidateExpandPath(Expression input, DataServiceContext con
 
                 throw new NotSupportedException(Strings.ALinq_InvalidExpressionInNavigationPath(input));
             }
+
+            /// <summary>
+            /// Checks whether the specified <paramref name="expr"/> is a valid aggregate expression.
+            /// </summary>
+            /// <param name="expr">The aggregate expression</param>
+            internal static void ValidateAggregateExpression(Expression expr)
+            {
+                MemberExpression memberExpr = StripTo<MemberExpression>(expr);
+                
+                // Disallow unsupported scenarios like d1.Prop.get_PropertyName() - e.g. d1.Prop.get_Length()
+                if (memberExpr == null)
+                {
+                    throw new NotSupportedException(Strings.ALinq_InvalidAggregateExpression(expr));
+                }
+
+                // Validate that property is aggregatable. Applies to CountDistinct() since Queryable 
+                // aggregation methods (Average, Sum, Min, Max) validate that the property is aggregatable
+                if (!PrimitiveType.IsKnownNullableType(memberExpr.Type))
+                {
+                    throw new NotSupportedException(Strings.ALinq_InvalidAggregateExpression(expr));
+                }
+
+                // Disallow unsupported scenarios like the following:
+                // - Average(d1 => d1.Prop.Length)
+                // - Average(d1 => d1.CollectionProp.Count)
+                MemberExpression parentExpr = StripTo<MemberExpression>(memberExpr.Expression);
+                if (parentExpr != null)
+                {
+                    if (PrimitiveType.IsKnownNullableType(parentExpr.Type))
+                    {
+                        throw new NotSupportedException(Strings.ALinq_InvalidAggregateExpression(expr));
+                    }
+
+                    Type collectionType = ClientTypeUtil.GetImplementationType(parentExpr.Type, typeof(ICollection<>));
+                    if (collectionType != null)
+                    {
+                        throw new NotSupportedException(Strings.ALinq_InvalidAggregateExpression(expr));
+                    }
+                }
+            }

```

Comment:
What is it that's not supported? Is it accessing the length of properties like strings or collections or is it accessing nested properties? is `d1.Prop.ChildProp` supported? and is `d1.Rectangle.Length` supported (assuming `Square` is not a collection)?

Reply:
@habbes Nested properties are supported. E.g. If you have a navigation property `Product`, you can group by property `ProductName` of `Product` type as follows:
```csharp
GroupBy(d1 => d1.Product.ProductName, ... )
```
What is not supported, is invoking properties of a known primitive type in the grouping expression. E.g. say `ProductName` in my example is of type `string`, it would be invalid to have
```csharp
GroupBy(d1 => d1.Product.ProductName.Length, ... )
```
We throw a `NotSupportedExpression` if that is used in the `GroupBy` `keySelector` expression

Note that this validation does not stop you from doing this:
```csharp
GroupBy(d1 => d1.Product.ProductName, (d1, d2) => new { ProductNameLength = d1.Length, ... });
```
Invoking the `Length` property (or any other property or method of the known primitive type) is just disallowed in the GroupBy `keySelector` expression.

Reply:
Maybe the comment or error-message should be more clear that the scenario being guarded against is accessing properties of known primitive types.